src/compiler/types.ts

```diff
@@ -935,27 +927,43 @@ module ts {
         getCurrentDirectory(): string;
     }
 
+    export interface WriteFileCallback {
+        (fileName: string, data: string, writeByteOrderMark: boolean, onError?: (message: string) => void): void;
+    }
+
     export interface Program extends ScriptReferenceHost {
         getSourceFiles(): SourceFile[];
-        getCompilerHost(): CompilerHost;
 
-        getDiagnostics(sourceFile?: SourceFile): Diagnostic[];
+        /**
+         * Emits the javascript and declaration files.  If targetSourceFile is not specified, then 
+         * the javascript and declaration files will be produced for all the files in this program.
+         * If targetSourceFile is specified, then only the javascript and declaration for that
+         * specific file will be generated.  
+         *
+         * If writeFile is not specified then the writeFile callback from the compiler host will be
+         * used for writing the javascript and declaration files.  Otherwise, the writeFile parameter
+         * will be invoked when writing the javascript and declaration files.
+         */
+        emit(targetSourceFile?: SourceFile, writeFile?: WriteFileCallback): EmitResult;
+
+        getSyntacticDiagnostics(sourceFile?: SourceFile): Diagnostic[];
         getGlobalDiagnostics(): Diagnostic[];
-        getDeclarationDiagnostics(sourceFile: SourceFile): Diagnostic[];
+        getSemanticDiagnostics(sourceFile?: SourceFile): Diagnostic[];
+        getDeclarationDiagnostics(sourceFile?: SourceFile): Diagnostic[];

```

Comment:
We do care.  Both tsc.ts and the services layer care about these different classifications.

1) The LS wants to know about syntactic errors so it can report them right away to the user so the user can easily fix their syntactic mistakes.  This is a concept of the VS diagnostics subsystem that is deeply threaded through everywhere.

2) The LS also needs to know about Declaration diagnostics.  We want to tell people (who have the -d flag) *up front* that there are issues that would cause Declaration problems.  As such, we want to report them when we report normal semantic errors.  However, we don't want to merge Declaration and Semantic, because to get Declaration errors you essentially must do an emit.  And htat would cause tsc to do two emits.

3) tsc wants to know about syntactic and semantic errors so it can report syntax, and then only report semantics if no syntax errors happened.  At hte command line this is valuable for preventing cascading errors that can make it difficult to discover the root cause.

4) Both the LS and tsc need to know about global errors so that they can be reported.

As such, these four methods are the smallest number i could come up with that met all our needs.  This system is beneficial over the old one in that:
a) there is only one place to go to get diagnostics.
b) diagnostics are cleanly separated.  You can't get the same diagnostic from different calls.
c) diagnostics should not be missed.  For example, we had bugs where some code was checking program.getGlobalDiagnostics but was missing typeChecker.getGlobalDiagnostics.  
d) Consumers who don't need this level of granularity can just call getPreEmitDiagnostics to get all relevant diagnostics.